### ISSUE TITLE
Mini Project: Lazy<T> filter

### ISSUE THREAD

**Issue by jeromepui** [nus-cs2030/2122-s1#164]
## Description 

`Lazy<T> filter(Predicate<? super T> predicate)` applies the given predicate lazily to test the content of the calling Lazy instance. If the predicate returns true, the content is retained, otherwise, the content becomes null. **If the content is already null, then the calling object is returned.**

```
jshell> Lazy.ofNullable(null)
$.. ==> Lazy[null]
jshell> Lazy.ofNullable(null).get()
$.. ==> Optional.empty
jshell> Lazy.ofNullable(null).map(x -> 1)
$.. ==> Lazy[?]
jshell> Lazy.ofNullable(null).filter(x -> true)
$.. ==> Lazy[?]
jshell> Lazy.ofNullable(null).filter(x -> false)
$.. ==> Lazy[?]
```

Shouldn't the last two test cases return `Lazy[null]`? Since the content is already null. Am I supposed to check at the start of the function that the supplier contains null and then return `this`?

**Comment by rui-han-crh** [comment]
It's because filter is not a terminating function, at this point in time, we are not yet aware if the filter result will or will not be null. It will only show null if `get()` is called upon to it and then we ask what the Lazy object looks like now.

Basically `filter` is still lazy, save the evaluation for later and check if it is null or not when a terminating `get` is called, then assign the evaluated value to the cached result so subsequent prints will show the result after evaluation. Any Lazy object prior to evaluation prints "?" unless it was already given a cached result in the first place.

**Comment by jeromepui** [comment]
I see, thanks! So does it want me to return the calling object if the evaluated result still gives null?

**Comment by rui-han-crh** [comment]
If the result is null, I believe the return type the test cases want is `Optional.empty`.
For example,
```
Lazy t = Lazy.ofNullable(null).filter(x -> false)        //returns Lazy[?]
t.get()                                                                      //returns Optional.empty
t                                                                              //shows Lazy[null]
```

**Comment by jeromepui** [comment]
Oh I meant if the result was null, does filter return the original `Lazy.<T>ofNullable(null)`, i.e. , `return this`, or does it return a new `Lazy.<T>ofNullable(null)`?

**Comment by rui-han-crh** [comment]
Oh that's interesting, I interpreted that sentence as being conducted after the evaluation of the predicate, so I gave this task of returning null or anything else to the get() function. But I suppose it can be seen as telling the filter function to check if the cache is null and return the calling object. But now I see your point that `Lazy[null] `would have been returned.

I only checked for null after the lazy filtering of predicate, using the `Optional.orElse(null)`
